R/ui.R

```diff
@@ -0,0 +1,186 @@
+        api_url <- sub(paste0("__(", paste0(names(.globals$UIs$mount), collapse = "|"),")__/$"), "", api_url)
+      }
+    }
+
+    utils::modifyList(list(servers = list(list(url = api_url))), spec)
+
+  }
+  # http://spec.openapis.org/oas/v3.0.3#document-structure
+  # "It is RECOMMENDED that the root OpenAPI document be named: openapi.json"
+  pr$handle("GET", "/openapi.json", openapi_fun, serializer = serializer_unboxed_json())
+  pr$handle("GET", "/openapi.yaml", openapi_fun, serializer = serializer_yaml())
+
+  invisible()
+
+}
+
+#' Mount OpenAPI Specification to a plumber router
+#' @noRd
+unmount_openapi <- function(pr) {
+
+  pr$removeHandle("GET", "/openapi.json")
+  pr$removeHandle("GET", "/openapi.yaml")
+  invisible()
+
+}
+
+#' Add UI for plumber to use
+#' @param ui A list of that plumber can use to mount
+#' a UI.
+#' @details [add_ui()] is used by package like `swagger` and `redoc`.
+#' When you load those package, it calls [add_ui()] to provide a user
+#' interface that can interpret your plumber OpenAPI Specifications.
+#'
+#' `ui` list expects the following values
+#' \describe{
+#' \item{package}{Name of the package required for the UI.}
+#' \item{name}{Name of the UI.}
+#' \item{index}{A function that returns the HTML content of the landing page of the UI.}
+#' \item{static}{A function that returns the path to the assets (images, javascript, css, fonts) the UI will use.}
+#' }
+#' @export
+add_ui <- function(ui) {
+
+  stopifnot(is.list(ui))
+  stopifnot(is.character(ui$package) && length(ui$package) == 1L)
+
+  if (!requireNamespace(ui$package, quietly = TRUE)) {
+    stop(ui$package, " must be installed for the ", ui$name," UI to be displayed")
+  }
+
+  stopifnot(is.character(ui$name) && length(ui$name) == 1L)
+  stopifnot(is.function(ui$static))
+  stopifnot(is.function(ui$index))
+  ui_root <- paste0("/__", ui$name, "__/")
+  ui_path <- paste0(ui_root, c("index.html", ""))
+
+  mount_ui_func <- function(pr, api_url, ...) {
+
+    ui_url <- paste0(api_url, ui_root)
+
+    ui_index <- function() {
+      ui$index(...)
+    }
+    for (path in ui_path) {
+      pr$handle("GET", path, ui_index,  serializer = serializer_html())
+    }
+    pr$mount(ui_root, PlumberStatic$new(ui$static(...)))
+    return(ui_url)
+  }
+  unmount_ui_func <- function(pr) {
+    for (path in ui_path) {
+      pr$removeHandle("GET", path)
+    }
+    pr$unmount(ui_root)
+    invisible()
+  }
+
+  .globals$UIs$mount[[ui$name]] <- mount_ui_func
+  .globals$UIs$unmount[[ui$name]] <- unmount_ui_func
+
+  invisible()
+}
+
+swagger_ui <- list(
+  package = "swagger",
+  name = "swagger",
+  index = function(version = "3", ...) {
+    swagger::swagger_spec(
+      api_path = 'window.location.origin + window.location.pathname.replace(/\\(__swagger__\\\\/|__swagger__\\\\/index.html\\)$/, "") + "openapi.json"',
+      version = version
+    )
+  },
+  static = function(version = "3", ...) {
+    swagger::swagger_path(version)
+  }
+)
+
+add_ui(swagger_ui)
```

Comment:
Can this be moved into the `.onLoad()` similar to how we did the parsers and serializers

Reply:
Yes, this will have to end up in swager package at some point.